src/hb-coretext-font.cc

```diff
@@ -0,0 +1,355 @@
+
+  if (len == -1)
+    len = strlen (name);
+
+  CFStringRef cf_name = CFStringCreateWithBytes (kCFAllocatorDefault,
+						 (const UInt8 *) name, len,
+						 kCFStringEncodingUTF8, false);
+  CGGlyph cg_glyph = CTFontGetGlyphWithName (ct_font, cf_name);
+  *glyph = cg_glyph;
+
+  CFRelease (cf_name);
+
+  // TODO Return true for .notdef; hb-ft does that.
+
+  return cg_glyph != 0;
+}
+
+
+static inline void free_static_coretext_funcs ();
+
+static struct hb_coretext_font_funcs_lazy_loader_t : hb_font_funcs_lazy_loader_t<hb_coretext_font_funcs_lazy_loader_t>
+{
+  static hb_font_funcs_t *create ()
+  {
+    hb_font_funcs_t *funcs = hb_font_funcs_create ();
+
+    hb_font_funcs_set_nominal_glyph_func (funcs, hb_coretext_get_nominal_glyph, nullptr, nullptr);
+    hb_font_funcs_set_nominal_glyphs_func (funcs, hb_coretext_get_nominal_glyphs, nullptr, nullptr);
+    hb_font_funcs_set_variation_glyph_func (funcs, hb_coretext_get_variation_glyph, nullptr, nullptr);
+
+    hb_font_funcs_set_font_h_extents_func (funcs, hb_coretext_get_font_h_extents, nullptr, nullptr);
+    hb_font_funcs_set_glyph_h_advances_func (funcs, hb_coretext_get_glyph_h_advances, nullptr, nullptr);
+    //hb_font_funcs_set_glyph_h_origin_func (funcs, hb_coretext_get_glyph_h_origin, nullptr, nullptr);
+
+#ifndef HB_NO_VERTICAL
+    //hb_font_funcs_set_font_v_extents_func (funcs, hb_coretext_get_font_v_extents, nullptr, nullptr);
+    //hb_font_funcs_set_glyph_v_advances_func (funcs, hb_coretext_get_glyph_v_advances, nullptr, nullptr);
+    //hb_font_funcs_set_glyph_v_origin_func (funcs, hb_coretext_get_glyph_v_origin, nullptr, nullptr);
+#endif
+
+#ifndef HB_NO_DRAW
+    hb_font_funcs_set_draw_glyph_func (funcs, hb_coretext_draw_glyph, nullptr, nullptr);
+#endif
+
+    hb_font_funcs_set_glyph_extents_func (funcs, hb_coretext_get_glyph_extents, nullptr, nullptr);
+
+#ifndef HB_NO_OT_FONT_GLYPH_NAMES
+    //hb_font_funcs_set_glyph_name_func (funcs, hb_coretext_get_glyph_name, nullptr, nullptr);
+    hb_font_funcs_set_glyph_from_name_func (funcs, hb_coretext_get_glyph_from_name, nullptr, nullptr);
+#endif
+
+    hb_font_funcs_make_immutable (funcs);
+
+    hb_atexit (free_static_coretext_funcs);
+
+    return funcs;
+  }
+} static_coretext_funcs;
+
+static inline
+void free_static_coretext_funcs ()
+{
+  static_coretext_funcs.free_instance ();
+}
+
+static hb_font_funcs_t *
+_hb_coretext_get_font_funcs ()
+{
+  return static_coretext_funcs.get_unconst ();
+}
+
+
+/**
+ * hb_coretext_font_set_funcs:
+ * @font: #hb_font_t to work upon
+ *
+ * Configures the font-functions structure of the specified
+ * #hb_font_t font object to use CoreText font functions.
+ *
+ * In particular, you can use this function to configure an
+ * existing #hb_face_t face object for use with CoreText font
+ * functions even if that #hb_face_t face object was initially
+ * created with hb_face_create(), and therefore was not
+ * initially configured to use CoreText font functions.
```

Comment:
Is this actually true? I don’t see where the `CTFont` is created in this case, since `hb_coretext_font_get_ct_font()` does not create one, but only gets the `CTFont` tht was passed to `hb_coretext_font_create()`.

Reply:
I was also confused by it, but I *think* it's created on demand, by this code:
```
CTFontRef 
hb_coretext_font_get_ct_font (hb_font_t *font) 
{ 
  CTFontRef ct_font = (CTFontRef) (const void *) font->data.coretext; 
  return ct_font ? (CTFontRef) ct_font : nullptr; 
} 
```
That `data.coretext` is a `hb_shaper_lazy_loader_t` that will call `_hb_coretext_shaper_font_data_create` on demand. No?

Reply:
I see. That makes sense!